src/midi/midi_fluidsynth.cpp

```diff
@@ -0,0 +1,245 @@
+/*
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *
+ *  Copyright (C) 2002-2011  The DOSBox Team
+ *  Copyright (C) 2020       Nikos Chantziaras <realnc@gmail.com>
+ *  Copyright (C) 2020       The dosbox-staging team
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "midi_fluidsynth.h"
+
+#if C_FLUIDSYNTH
+
+#include <clocale>
+#include <string>
+
+#include "control.h"
+
+// Fluidsynth uses libinstpatch which as of v1.1.3 has the nasty behavior of
+// changing the locale under our nose. We remember and restore the current
+// locale whenever this could happen.
+//
+// https://github.com/realnc/dosbox-core/issues/7
+// https://github.com/swami/libinstpatch/issues/37
+//
+class LocaleGuard final {
+public:
+	~LocaleGuard() { setlocale(LC_ALL, locale_.c_str()); }
+
+private:
+	std::string locale_ = setlocale(LC_ALL, nullptr);
+};
+
+MidiHandlerFluidsynth MidiHandlerFluidsynth::instance;
+
+void init_fluid_dosbox_settings(Section_prop &secprop)
+{
+	constexpr auto when_idle = Property::Changeable::WhenIdle;
+
+	auto *str_prop = secprop.Add_string("fluid.soundfont", when_idle, "");
+	str_prop->Set_help("Soundfont to use with FluidSynth."
+	                   "One must be specified.");
+
+	auto *int_prop = secprop.Add_int("fluid.samplerate", when_idle, 44100);
+	int_prop->SetMinMax(8000, 96000);
+	int_prop->Set_help("Sample rate (Hz) Fluidsynth will render at. (min "
+	                   "8000, max 96000");
+
+	str_prop = secprop.Add_string("fluid.gain", when_idle, "0.4");
+	str_prop->Set_help("Fluidsynth volume gain. (min 0.0, max 10.0");
+
```

Comment:
This is probably best for a follow on PR, but we should be able to abstract this one away to and get the gain from the `mixer midi PERCENT` setting; or at least have the `mixer midi` setting also be able to adjust this setting.

This is because I'd rather Fluid be the one scaling the samples up or down because it closer to the source material and can presumably do a better job - especially given the polyphony nature of what it's doing (ie: mixing lots of voices into a stream), versus having dosbox's mixer do the scaling - which we know will simply clip and distort samples (not saying that's wrong or bad..)

**mixer.cpp:**
``` c++
static inline int16_t MIXER_CLIP(Bits SAMP)
{
        if (SAMP < MAX_AUDIO) {
                if (SAMP > MIN_AUDIO)
                        return SAMP;
                else return MIN_AUDIO;
        } else return MAX_AUDIO;
}
```

Reply:
Exactly; this PR does not allow users to change fluid's `synth.gain` so we could make it fit dosbox internals in subsequent PR (if it'll be even necessary).